_drafts/2023-06-30-kotlin-context-receivers.md

```diff
@@ -0,0 +1,517 @@
+
+Let's try to understand if context receivers are suitable for dependency injection with an example. Let's say we have a `JobsController` class that exposes jobs as JSON and uses a `Jobs` module to retrieve them. We can define it as follows:
+
+```kotlin
+context (Jobs, JsonScope<Job>, Logger)
+class JobController {
+    suspend fun findJobById(id: String): String {
+        log(Level.INFO, "Searching job with id $id")
+        val jobId = JobId(id.toLong())
+        return findById(jobId)?.let {
+            log(Level.INFO, "Job with id $id found")
+            return it.toJson()
+        } ?: "No job found with id $id"
+    }
+}
+```
+
+We must provide the three required contexts to use the `JobController` class. We can do it using the `with` scope function, as we previously did:
+
+```kotlin
+suspend fun main() {
+    with(jobJsonScope) {
+        with(consoleLogger) {
+            with(LiveJobs()) {
+                JobController().findJobById("1").also(::println)
+            }
+        }
+    }
+}
+```
+
+We can see that the `JobController` class has three context receivers: `Jobs`, `JsonScope<Job>`, and `Logger`. Inside the `findJobById` method, the contexts are accessed without specification. The `log` method and the `findById` function are called as part of the `JobController` class.
+
+The above code makes it unclear which method belongs to which class. Who owns the function `findById` or the function `log`? In general, **implicit function resolution is harder to read and understand** and, thus, to maintain. Moreover, we can't avoid name clashes when using multiple contexts.
+
+We can change and make it more explicit by using the `@` notation to access the context receivers. For example, we can rewrite the `findJobById` method as follows:
+
+```kotlin
+context (Jobs, JsonScope<Job>, Logger)
+class JobController {
+    suspend fun findJobById(id: String): String {
+        this@Logger.log(Level.INFO, "Searching job with id $id")
+        val jobId = JobId(id.toLong())
+        return this@Jobs.findById(jobId)?.let {
+            this@Logger.log(Level.INFO, "Job with id $id found")
+            return it.toJson()
+        } ?: "No job found with id $id"
+    }
+}
+```
+
+However, the notation is very verbose and makes the code less readable. Moreover, we must ensure that a developer uses it.
+
+In Scala, we had a very close problem with the [Tagless Final encoding](https://blog.rockthejvm.com/tagless-final/) pattern. In the past, many Scala developers started to use the pattern to implement dependency injection. Using a similar approach to context receivers, Scala allows us to define type constraints in the type parameters definition. The `JobController` class would look like the following in Scala:
```

Comment:
As a Scala dev, I love seeing this analogy. I'm not super sure if Kotlin folks will take a lot of value out of this section though. Need to think about it

Reply:
I let you think about this. IMHO, Kotlin devs should learn from mistakes made by other languages.

Reply:
I know, it's just that it adds mental burden to non-Scala devs to understand TF first (which is non-trivial even for Scala devs)